onnxruntime/core/mlas/lib/pooling.cpp

```diff
@@ -1317,35 +1340,12 @@ Return Value:
         }
 
         if (ReductionBufferRemaining >= int64_t(WorkBlock.InputShape[Dimensions - 1])) {
-            PoolKernelRoutine = MlasPoolVectorKernels[PoolingKind][Dimensions - 2];
+            WorkBlock.PoolKernelRoutine = MlasPoolVectorKernels[PoolingKind][Dimensions - 2];
         }
     }
 
-#ifdef MLAS_NO_ONNXRUNTIME_THREADPOOL
-    MLAS_UNREFERENCED_PARAMETER(ThreadPool);
-    //
-    // Execute the pooling kernel routine.
-    //
-
-#if defined(_OPENMP)
-
-#pragma omp parallel for
-    for (int64_t c = 0; c < int64_t(TotalChannelCount); c++) {
-      PoolKernelRoutine(&WorkBlock, 1, Input + c * InputSize, Output + c * OutputSize);
+    if (WorkBlock.TargetThreadCount >= TotalChannelCount) {
+      WorkBlock.TargetThreadCount = TotalChannelCount;
     }
-
-#else
-
-    PoolKernelRoutine(&WorkBlock, TotalChannelCount, Input, Output);
-
-#endif
-#else
-    //
-    // Use an external thread pool if one is provided.
-    // TODO: change to use MlasExecuteThreaded
-    onnxruntime::concurrency::ThreadPool::TryBatchParallelFor(ThreadPool, static_cast<int32_t>(TotalChannelCount), [&](int32_t c) {
-      PoolKernelRoutine(&WorkBlock, 1, Input + c * InputSize, Output + c * OutputSize);
-    });
-    return;
-#endif    
+    MlasExecuteThreaded(MlasPoolThreaded, &WorkBlock, int32_t(WorkBlock.TargetThreadCount), ThreadPool);
```

Comment:
How does the performance of this compare for various CNNs where you weren't already building with openmp? Does this regress for Windows ML that wasn't already carrying this around?

Reply:
Sorry I didn't understand your question. When OpenMP is disabled, I guess there isn't any big difference between  ThreadPool::TryBatchParallelFor and MlasExecuteThreaded ?

Reply:
I guess I missed or ignored the change you made in January that added TryBatchParallelFor. The way it was before that, the non-openmp build didn't do any threading. The new change is overly aggressive with threading. I'll re-evaluate the change given the current state, but again it should be based on complexity of the problem, not just using the channel count. This maybe isn't noticed as much now because x64 CNNs go through the NCHWc path.